Nullable Annotations Guidance (#40940)
dotnet/roslyn: docs/contributing/Nullable Annotations.md

# Nullable Annotations

This document describes how nullable annotations should be approached in the 
Roslyn code base. The default is to simply follow [the same guidance](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/api-guidelines/nullability.md)
as the [dotnet/runtime](github.com/dotnet/runtime) repository. This document
serves to detail the places where the guidance differs for Roslyn and 
re-emphasize rules that come up frequently.

## Annotation Guidance 

- **DO** annotate all new code
- **DO** use the null suppression operator in places where a method body has
ensured a value cannot be `null` but the compiler is unable to track it. 
Example:
```cs
struct SyntaxTrivia {
    void Example() {    
        if (this.HasStructure) {

>             Use(this.GetStructure()!)

Reviewer: ```suggestion
            Use(this.GetStructure()!);
```

        }
    }
}
```
- **DO** use `Assert` to validate state that is passed into the method
```cs
void M1(SyntaxNode node) {
    RoslynDebug.Assert(node.Parent is object);
    ...
    M2(node.Parent);
}
```
- **AVOID** annotating and refactoring code in the same commit. This can
significantly increase the cost of reviewing. Prefer instead to separate 
annotations and significant refactorings into separate commits. Small changes,
and modernizing style (change `default(SyntaxToken)` to `default`) are

> acceptable.

Reviewer: One example is renaming a parameter to remove the `Opt` suffix. I found it's often cleaner to avoid the rename initially, and then rename it as a future commit.

Author: The `Opt` suffix is borderline for me. Prefer we do it as a separate commit but not so much that I put up a fight over it.

---
In reply to: [366078764](https://github.com/dotnet/roslyn/pull/40940#discussion_r366078764) [](ancestors = 366078764)

Reviewer 2: The one thing I found icky about doing Opt in a separate commit was having to go back and find all the places you have to fix up; it's easy to fix both the annotation and name at once.

- **DO NOT** be concerned with the validity of annotations after an object has 

> been Dispose'd or otherwise entered into an invalid state. For example

Reviewer: ```suggestion
been disposed or otherwise entered into an invalid state. For example
```

manipulating a pooled object after it's been freed. This is a violation of the 
objects contract and annotations aren't meaningful here.

> - **DO NOT** be concerned with the validity of an enumerator's Current 

Reviewer: ```suggestion
- **DO NOT** be concerned with the validity of an enumerator's `Current` 
```

> annotations when used before MoveNext has been called or after MoveNext has 

Reviewer: ```suggestion
annotations when used before `MoveNext` has been called or after `MoveNext` has 
```

> returned false.  As with Dispose, such use is considered invalid, and we don't

Reviewer: ```suggestion
returned false.  As with `Dispose`, such use is considered invalid, and we don't
```

> want to harm the correct consumption of Current by catering to incorrect

Reviewer: ```suggestion
want to harm the correct consumption of `Current` by catering to incorrect
```

consumption.

> ## Breaking Change Guidance

Reviewer: ```suggestion
## Breaking Change Guidance

```

# Nullable Annotations

This document describes how nullable annotations should be approached in the 
Roslyn code base. The default is to simply follow [the same guidance](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/api-guidelines/nullability.md)
as the [dotnet/runtime](github.com/dotnet/runtime) repository. This document
serves to detail the places where the guidance differs for Roslyn and 
re-emphasize rules that come up frequently.

## Annotation Guidance 

- **DO** annotate all new code
- **DO** use the null suppression operator in places where a method body has
ensured a value cannot be `null` but the compiler is unable to track it. 
Example:
```cs
struct SyntaxTrivia {
    void Example() {    
        if (this.HasStructure) {
            Use(this.GetStructure()!)
        }
    }
}
```

> - **DO** use `Assert` to validate state that is passed into the method

Author: I'm not sure what you mean by "historically" here. Nullable annotations are very new.

I also don't see the reason why we should differ between compiler and IDE here. Can you give reasons why we should differ on this? Is there an IDE style we should be considering instead? 

Having different rules for different layers just makes a hard problem that much harder. Better to standardize and move forward together.

> - **DO** use `Assert` to validate state that is passed into the method

Reviewer: ```suggestion
- **CONSIDER** using `Assert` to validate state that is passed into the method if it avoids the need for suppressions later in the method
```

> - **DO** use `Assert` to validate state that is passed into the method

Reviewer 2: Historically we didn't use Assert for any null checking, we always did Contract.ThrowIfNull(). This was the policy that predates null annotations and honestly goes back many many years. More or less the logic was the same reason Go doesn't offer asserts: because they're not in retail code, it means you end up looking at Watsons that didn't behave how the developer imagined.

> Having different rules for different layers just makes a hard problem that much harder. Better to standardize and move forward together.

Agreed -- this something we never agreed on historically, this document just reminds us of that. :smile:

> - **DO** use `Assert` to validate state that is passed into the method

Author: Ah okay. I have no objection to `Contract.ThrowIfNull`. I was more getting that wee should use a form fo explicit validation, Assert, ThrowIfNull, for external state. Wasn't trying to mandate `Assert` specifically. Let me update that language.

---
In reply to: [366084477](https://github.com/dotnet/roslyn/pull/40940#discussion_r366084477) [](ancestors = 366084477)

```cs
void M1(SyntaxNode node) {
    RoslynDebug.Assert(node.Parent is object);
    ...
    M2(node.Parent);
}
```
- **AVOID** annotating and refactoring code in the same commit. This can
significantly increase the cost of reviewing. Prefer instead to separate 
annotations and significant refactorings into separate commits. Small changes,
and modernizing style (change `default(SyntaxToken)` to `default`) are
acceptable.
- **DO NOT** be concerned with the validity of annotations after an object has 
been Dispose'd or otherwise entered into an invalid state. For example
manipulating a pooled object after it's been freed. This is a violation of the 
objects contract and annotations aren't meaningful here.
- **DO NOT** be concerned with the validity of an enumerator's Current 
annotations when used before MoveNext has been called or after MoveNext has 
returned false.  As with Dispose, such use is considered invalid, and we don't
want to harm the correct consumption of Current by catering to incorrect
consumption.

## Breaking Change Guidance
At this point in time nullable annotations are not considered as a part of our 

> compatibility bar. That is nullable annotations can change as needed or desired

Reviewer 2: I wouldn't make such a strong statement here -- some parts of the Workspaces layer we've already annotated and had discussions; further changes there I'd expect long conversations on that would probably include back-compat conversations. I don't want somebody interpreting this as "I can make changes willy-nilly and it's fine."

Reviewer 2: Put another way: this feels like the compiler and IDE might be at different levels of confidence of our annotations in some places, and different parts of the code might be at different levels of breaking change guidance accordingly.

Author: Why is the IDE so confident their annotations are correct even though virtually all the code they depend on is oblivious? That includes both the code in Roslyn and in the SDK. Believing the APIs are done in the face of only oblivious code seems premature.

Also if you all are considering annotations to be a part of the compat bar did you clear that with @333fred (since he owns API reviews). 

I see no reason to commit to a compat bar yet. Particularly because the code base is in flux. A few months from now when we're much further along with the core annotations I'll feel significantly better about attaching compat.

Reviewer: I am in favor of the current warning. While we don't _want_ to make these changes, we are wholly reserving the right to do so.

Author: To be clear: if the IDE wants to apply more structiny to their changes I'm not against that. At the same time I think it's premature to commit to compatibility with our customer base.

Reviewer 2: > Why is the IDE so confident their annotations are correct even though virtually all the code they depend on is oblivious?

Some of the code we've already annotated and in doing so annotated everything they call into.   No compiler API nor framework change at this point would result in different analysis.

> To be clear: if the IDE wants to apply more structiny to their changes I'm not against that.

I'm not saying I'd want to commit 100% to it's not changing, and I'd say we reserve the right to make a breaking change like we do at any time (because "we don't make breaking changes" is never true in the real world). I just would expect additional scrutiny at this point in the already annotated parts of the IDE layer, because it means there's something far deeper afoot and we either misannotated something (and we need to understand what went wrong) or we are potentially changing a contract.

Reviewer 3: I do think there's some difference between the syntax APIs and the rest of our APIs here: if the language adds a new syntactic form that makes some element optional, well, the contract needs to change. The SemanticModel, though, likely shouldn't make an arbitrary decision to start returning `null` from some `GetTypeInfo` call where we never return `null` today.

Reviewer 2: And perhaps another way to make the distinction I had earlier: I fully expect as the compiler annotates say syntax layer that the IDE _implementation_ will change accordingly. I'm sure you're going to discover at least one syntax node property somewhere that can be null that we didn't expect and we get to add a bunch of null checks. But that won't change, say, the Rename APIs we have suddenly exposing null somewhere. Those APIs are so high level that just no syntax or semantics or symbols are exposed from them in any meaningful way that our signatures will change. If they're changing that implies it's like what @333fred is saying: it implies we actually did make an arbitrary decision to return null when we didn't before. Or we now went onto record saying we weren't, and there's perhaps impact in changing that.

Author: I'm still lost on how to adjustthis section. The IDE team is both stating "we want to keep compat" and "we want to change compat" so which is it? The IDE seems to want to have it both ways here. 

Also if the IDE team is concerned with compat how are you tracking it? Have you enabled tracking in your API files? Do you just scrutize PRs and attempt to spot compat changes manually? 

What I want to get out of this section is the following:

1. We're not complete so expect changes if you use `#nullable` and consume Roslyn as an API. Such changes in much of the code will occur without any notice: no entry in a breaking change log, no tracking in API files, etc ...
1. We will get to a place where we are complete enough that nullable changes will be a part of our compatibility bar. It will never be as high as the non-nullable portions of our API. That is if the language, or underlying editor, forces us to change nullable annotations a member we will do so. But we will not arbitrarily change nullable annotations at that point. 
1. Want to express how a customer viewing the source can tell when a library has reached a point that nullable annotations are a part of the compat bar.

What I specifically do **not** want to say:

> IDE and compiler have different policies here. 

That statement is not actionable by our customers and not enforceable. The terms IDE and Compiler are our terms. The granularity customers see is libraries. 

---
In reply to: [366109197](https://github.com/dotnet/roslyn/pull/40940#discussion_r366109197) [](ancestors = 366109197)

Author: Attempted to rework this a bit to get your points across. 

---
In reply to: [367127555](https://github.com/dotnet/roslyn/pull/40940#discussion_r367127555) [](ancestors = 367127555,366109197)

Reviewer 2: I love the phrasing now: it makes it clear that to start things aren't a compat bar, but at some point certain areas will get updated to be "done" or "good enough", and this document establishes itself as the place where that's done.

by the Roslyn team and it is not considered a compat change even if it
introduces new warnings into consumer code.

This is necessary as Roslyn is a very large code base and it will take a 
significant amount of time to annotate the code base. Even Public API 

> annotations only is a siginifcant amount of work and, without the 

Reviewer 3: ```suggestion
annotations only is a significant amount of work and, without the 
```

    ...
    M2(node.Parent);
}
```
- **AVOID** annotating and refactoring code in the same commit. This can
significantly increase the cost of reviewing. Prefer instead to separate 
annotations and significant refactorings into separate commits. Small changes,
and modernizing style (change `default(SyntaxToken)` to `default`) are
acceptable.
- **DO NOT** be concerned with the validity of annotations after an object has 
been disposed or otherwise entered into an invalid state. For example
manipulating a pooled object after it's been freed. This is a violation of the 
objects contract and annotations aren't meaningful here.
- **DO NOT** be concerned with the validity of an enumerator's `Current` 
annotations when used before `MoveNext` has been called or after `MoveNext` has 
returned false. As with `Dispose`, such use is considered invalid, and we don't
want to harm the correct consumption of `Current` by catering to incorrect
consumption.
- **DO** use the adjective `Require` when defining a member that exist only
to provide a non-null returning version of a method that returns a nullable 
reference type. For example: `GetRequiredSemanticModel` model returns a non-null 

> value from `GetSemanticModel`

Reviewer 2: Is this for internal methods only or future public APIs that parallel legacy null-returning API? The IDE was doing this with internal methods but we were somewhat hedging on whether we wanted to make those public. We borrowed the name from some methods (GetRequiredLanguageService) that are very old and so on one hand liked the consistency, but weren't sure if that was following a consistency that wasn't really intended at the time. Those methods predate Roslyn 1.0 and thus establishing a naming convention for methods wasn't in the purview of that!

Did the framework do anything here?
